NRLMMD-GEOIPS/geoips
# 474 update cli for breaking change from product names

NOTE: A huge amount of files changed will be removed once #444  and #465 have been merged. This PR is a sub-branch of #465 merged into the changes that were made in #433. DO NOT MERGE until both #444 and #465 have been merged into main.

# Reviewer Checklist

* [x] NO REQUIRED ***unit tests*** (Added in CLI V2 PR #465)
* [x] NO REQUIRED ***integration tests*** 
* [x] Required ***updates to other repos*** complete (See NRLMMD-GEOIPS/data_fusion#29 NRLMMD-GEOIPS/data_fusion#31)

# Related Issues
fixes NRLMMD-GEOIPS/geoips#474
fixes NRLMMD-GEOIPS/geoips#423
fixes NRLMMD-GEOIPS/geoips#428

# Testing Instructions
Run ``geoips test unit-test commandline`` to run all CLI Unit Tests and you can also run ``./geoips/tests/integration_tests/full_test.sh`` to make sure everything else is working as expected.

# Summary

Create products for each possible product family handled by single_source procflow
----------------------------------------------------------------------------------

This update adds a new product ``all_product_families.yaml`` to the products interface.
This file provides every possible product family to test against ``single_source.py``
procflow plugin, as we need to understand which families work, and the process these
families follow throughout that code in order to implement a new procflow down the road,
called ``order_based.py`` procflow. By understanding the paths that these families take,
we set ourselves up for success in implementing that new procflow.

Files Added:

    - added: geoips/plugins/yaml/products/all_product_families.yaml
    - added:   tests/scripts/order_based_all_product_families/alg.abi.infrared.sh
    - added:   tests/scripts/order_based_all_product_families/alg_cmap.abi.infrared.sh
    - added:   tests/scripts/order_based_all_product_families/alg_int_cmap.abi.infrared.sh
    - added:   tests/scripts/order_based_all_product_families/int.abi.infrared.sh
    - added:   tests/scripts/order_based_all_product_families/int_alg.abi.infrared.sh
    - added:   tests/scripts/order_based_all_product_families/int_alg_cmap.abi.infrared.sh
    - added:   tests/scripts/order_based_all_product_families/sect_xdict_out_form.smap-spd.wind.sh
    - added:   tests/scripts/order_based_all_product_families/unmodified_prd.smap-spd.winds.sh
    - added:   tests/scripts/order_based_all_product_families/unsect_xdict_area_out_form.smap-spd.winds.sh
    - added:   tests/scripts/order_based_all_product_families/unsect_xdict_out_form.smap-spd.winds.sh
    - added:   tests/scripts/order_based_all_product_families/xdict_out_form.smap-spd.winds.sh

Note: See Bux Fix for duplicate source_names in product registry *GEOIPS#428*, for more
information on bug fixes made on this branch as well.

Fix product sub-plugins duplicate source name error
---------------------------------------------------

Currently, when the plugin registry is built, the key names associated with the
sub-products contained in a yaml file (for instance lets say smap-spd.yaml), are
gathered via the sub-products source name rather than the name of the product plugin
they reside in. In this case, all product plugins under the key name smap-spd, are
placed in that location because of their source name, not the attribute name, found
at the top level of smap-spd.yaml. This could cause issues with other products, as
found in issue *GEOIPS#423: 2024-1-17, create products for each possible product family*

This updates fixes this bug by changing the naming convention of product plugins stored
within the plugin registry. Instead of products -- <source_name> -- sub_product_name,
the keys are now formatted products -- <source_name>.<sub_product_name>, and duplicates
are checked on that level.

Files modified (excluding files which were modified for black formatting):

    -  modified: geoips/create_plugin_registries.py
    -  modified: geoips/interfaces/base.py
    -  modified: geoips/interfaces/yaml_based/products.py
    -  modified: geoips/dev/output_config.py
    -  modified: geoips/plugins/modules/filename_formatters/tc_file_naming.py
    -  modified: geoips/plugins/modules/output_formatters/imagery_annotated.py
    -  modified: geoips/plugins/modules/output_formatters/metadata_tc.py
    -  modified: geoips/plugins/modules/procflows/config_based.py
    -  modified: geoips/plugins/modules/procflows/single_source.py

Updated GeoIPS CLI to fix breaking changes introduced by all_product_families
-----------------------------------------------------------------------------

While no breaking changes have been introduced yet, Branch #423, PR #433 Test All
Product Families introduced changes to how product plugins are stored in the registry,
which would break some portions of the CLI. Specifically, product plugins are no longer
registered via a (source_name, product_name) tuple combination. They are still accessed
via products.get_plugin(source_name, product_name), however under the hood they are
registered as unique names via ``source_name.product_name``. This was added to prevent
errors that possibly could occur if different products refered to the same source_name
in different files. This broke some functionality of the CLI in ``geoips_command.py``
and ``geiops_get.py``.

Files Modified for Updating the CLI:

    modified: geoips/commandline/geoips_command.py
    modified: geoips/commandline/geoips_get.py
